execute/observation.go

```diff
@@ -233,12 +235,43 @@ func (p *Plugin) getCommitReportsObservation(
 		}
 	}
 
+	// Update the earliest unexecuted root based on remaining reports
+	combinedReports := p.buildCombinedReports(groupedCommits, fullyExecutedUnfinalized)
+	p.commitRootsCache.UpdateEarliestUnexecutedRoot(combinedReports)
+
 	observation.CommitReports = groupedCommits
 
 	// TODO: truncate grouped to a maximum observation size?
 	return observation, nil
 }
 
+// buildCombinedReports creates a combined map for updating the earliest unexecuted root
+func (p *Plugin) buildCombinedReports(
```

Comment:
nit: can be a free function

```suggestion
// buildCombinedReports creates a combined map for updating the earliest unexecuted root
func buildCombinedReports(
```